blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/credential/CredentialsUtils.java

```diff
@@ -0,0 +1,180 @@
+            Optional<Domain> d = Iterables.tryFind(store.getDomains(), new Predicate<Domain>() {
+                @Override
+                public boolean apply(@Nullable Domain input) {
+                    return input != null && Iterables.tryFind(store.getCredentials(input), new Predicate<Credentials>() {
+                        @Override
+                        public boolean apply(@Nullable Credentials input) {
+                            return (input != null && input instanceof IdCredentials) && ((IdCredentials) input).getId().equals(credentialId);
+                        }
+                    }).isPresent();
+
+                }
+            });
+            if (d.isPresent()) {
+                return d.get();
+            }
+        }
+        return null;
+    }
+
+    public static @CheckForNull <C extends Credentials> C  findCredential(@Nonnull String id, @Nonnull Class<C> type){
+        if(User.current() == null){
+            throw new ServiceException.UnauthorizedException("No authenticated user found. Please login");
+        }
+
+        return CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(
+                        type,
+                        Jenkins.getInstance(),
+                        Jenkins.getAuthentication(),
+                        Collections.<DomainRequirement>emptyList()),
+                CredentialsMatchers.allOf(CredentialsMatchers.withId(id))
+        );
+    }
+
+    public static @CheckForNull <C extends Credentials> C findCredential(@Nonnull String credentialId, @Nullable String uri, @Nonnull Class<C> type){
+        return CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(
+                        type,
+                        Jenkins.getInstance(),
+                        Jenkins.getAuthentication(),
+                        URIRequirementBuilder.fromUri(uri).build()),
+                CredentialsMatchers.allOf(CredentialsMatchers.withId(credentialId))
+        );
+    }
+
+    private static @CheckForNull CredentialsStore findUserStoreFirstOrNull(User user){
+        for(CredentialsStore s: CredentialsProvider.lookupStores(user)){
+            if(s.hasPermission(CredentialsProvider.CREATE) && s.hasPermission(CredentialsProvider.UPDATE)){
+                return s;
+            }
+        }
+        return null;
+    }
+
+    private static @CheckForNull Iterable<CredentialsStore> findUserStores(User user){
+        return Iterables.filter(CredentialsProvider.lookupStores(user), new Predicate<CredentialsStore>() {
+            @Override
+            public boolean apply(@Nullable CredentialsStore s) {
+                return s!= null && s.hasPermission(CredentialsProvider.CREATE) && s.hasPermission(CredentialsProvider.UPDATE);
+            }
+        });
+
+    }
+
+    private static List<DomainSpecification> generateDomainSpecifications(@Nullable URI uri){
+        if (uri == null) {
```

Comment:
You are aware of https://github.com/jenkinsci/credentials-plugin/blob/master/src/main/java/com/cloudbees/plugins/credentials/domains/URIRequirementBuilder.java